### PR TITLE
Added Touch ID support for unlocking password locked databases

### PR DESCRIPTION
**Pull request by L4grange**
What changed:

If the the device has TouchID support, ask the user if they want to save their DB password in the Keychain. This happens when the user changes the Master DB Password.

If they say yes, the DB password is saved in the keychain and the DB name is saved in a list of "Touch ID enabled" databases.

When MacPass asks for a locked DB's password, if the DB is in the "TouchID enabled" databases, offer to unlocked with TouchID

If finger authentication is successful, get the password from the keychain and unlock the DB. (otherwise fallback to normal password input)

P.S.: If the user changes the master password and says no to the "Use TouchID dialog", then any previously stored password for that database are removed.

Suggestions/corrections are very welcome! (Especially on where to put the list of TouchID Enabled databases, and to fix the TouchID window to only pop up when the app is focused)

### CODE REVIEW

**Comment by mstarke** [comment]
Nice work. I'll review the code as soon as i find the time. Love the contribution!

File: MacPass/MPPasswordEditWindowController.m [hunk 1]
```diff
@@ -136,6 +143,60 @@ - (IBAction)generateKey:(id)sender {
...
 
+-(void) _askForTouchID:(NSString*)password {
+	NSError *authError = nil;
+	LAContext *myContext = [LAContext new];
+	if ([myContext canEvaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics error:&authError]) {
+
+		NSAlert *alert = [NSAlert new];
+		[alert addButtonWithTitle:@"Yes"];
+		[alert addButtonWithTitle:@"No"];
+		[alert setMessageText:@"Use TouchID for unlocking this database?"];
```
Comment (mstarke) [inline MacPass/MPPasswordEditWindowController.m]:
````obj-c
alert.messageText = NSLocalizedString(@"ALERT_TOUCH_ID_MESSAGE",@"");
````

File: MacPass/MPPasswordEditWindowController.m [hunk 2]
```diff
@@ -136,6 +143,60 @@ - (IBAction)generateKey:(id)sender {
...
+		MPDocument *document = self.document;
+		NSString *dbName = [document displayName];
+
+		KeychainPasswordItem *passwordItem = [[KeychainPasswordItem alloc] initWithService:@"MacPass" account:dbName accessGroup:nil];
+
+		[passwordItem savePassword:password error:nil]; //Save the password in the keychain
+		[MPSettingsHelper addTouchIdEnabledDatabaseWithName:dbName]; //Add DB name in the list of Touch ID enabled databases
+		NSLog(@"Saved DB (%@) password in the keychain.", dbName);
+	}
+	@catch(NSException *exceptionm) {
```
Comment (mstarke) [inline MacPass/MPPasswordEditWindowController.m]: exception

**Comment by L4grange** [comment]
I've made some changes (localized all strings, replaced try/catch blocks with NSErrors) and I'm trying to implement the DB's root UUID instead of name, which I also think is the right way to go. (I was not aware of it's existence before you mentioned it).

The only problem I'm facing if that I need access to the UUID while the DB is locked so I can check if it is in the list of TouchID enabled databases. However, `[document.root.uuid UUIDString]` returns nil when the database is locked (I'm accessing it in the MPPasswordInputController). Any ideas on how to get around this? Thanks for your suggestions and for making this great KeePass client :)

P.S.: I'll look into the public DB metadata tomorrow, although if this metadata is written onto the DB file, and the user syncs the DB to multiple computers, then the "is TouchID enabled" for each database would not be valid on any other computer other than the one where the DB password is saved in the keychain (except if the keychain is synced with iCloud, but my implementation uses the local login keychain)

**Comment by mstarke** [comment]
You're right, the Root UUID is not available for unencrypted databases.
Another observation: You added the ability to add passwords to the keychain. This was requested as a standalone feature so a suggestion would be to generalise the approach.

TouchID devices can use a touchID encrypted keychain, other devices use the normal keychain if the user chooses to store it this way. This way you can even migrate keychain items.

The most secure way would be to use the secure enclave to encrypt a key that is then used to unlock the database but this only works for one device and defeats the purpose of sharing the database.

**Comment by mstarke** [comment]
You can add the changes you made to this pull request so it'll keep up to date!

**Comment by mstarke** [comment]
Also an aside - do you absolutely need to use swift? I'm a bit hesitant to ship the run time with every MacPass version just for a single file. Also this adds another layer of complexity since Swift changes are still frequent so MacPass needs to be updates all the time to keep up.

File: MacPass/MPPasswordInputController.m [hunk 3]
```diff
@@ -130,4 +140,73 @@ - (void)_showError:(NSError *)error {
   self.errorInfoTextField.hidden = NO;
 }
 
+- (IBAction)showTouchIdDialog:(id)sender {
+	[self _enableTouchID];
+}
+
+- (void) appFocused {
```
Comment (mstarke) [inline MacPass/MPPasswordInputController.m]: Please remove any unnecessary logging. I know it's nice to see that suff work ;-)

File: MacPass/MPPasswordInputController.m [hunk 4]
```diff
@@ -130,4 +140,73 @@ - (void)_showError:(NSError *)error {
...
+	}
+
+	LAContext *myContext = [LAContext new];
+	NSString *myLocalizedReasonString = @"unlock this database";
+
+	NSError *authError = nil;
+	//	if (#available(OSX 10.12, *)) {
+
+	if (floor(NSFoundationVersionNumber) > NSFoundationVersionNumber10_11_4) {
+		if ([myContext canEvaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics error:&authError]) {
```
Comment (mstarke) [inline MacPass/MPPasswordInputController.m]: Consider using a helper to wrap something like this like ```MPOSHelper.supportsTouchID``` or the like

File: MacPass/MPPasswordInputController.m [hunk 5]
```diff
@@ -130,4 +140,73 @@ - (void)_showError:(NSError *)error {
...
+				NSLog(@"Failure Reason: %@", authError.localizedFailureReason);
+			}
+		}
+	} else {
+		NSLog(@"TouchID is not supported.");
+	}
+}
+
+- (void) _getPasswordFromKeychain {
+	NSString *dbName = [self getDatabaseName];
```
Comment (mstarke) [inline MacPass/MPPasswordInputController.m]: ```self.databaseName````

File: MacPass/MPPasswordInputController.m [hunk 6]
```diff
@@ -130,4 +140,73 @@ - (void)_showError:(NSError *)error {
...
+	} else {
+		dispatch_sync(dispatch_get_main_queue(), ^{
+			_passwordTextField.stringValue = pass;
+			[self _submit:nil];
+		});
+	}
+
+}
+
+- (NSString*) getDatabaseName {
```
Comment (mstarke) [inline MacPass/MPPasswordInputController.m]: should be converted to ```@property```

File: MacPass/MPPasswordInputController.m [hunk 7]
```diff
@@ -130,4 +140,73 @@ - (void)_showError:(NSError *)error {
...
+	MPDocumentWindowController *documentWindow = self.windowController;
+	MPDocument *document = documentWindow.document;
+	return [document displayName];
```
Comment (mstarke) [inline MacPass/MPPasswordInputController.m]: ```document.displayname;```

File: MacPass/MPSettingsHelper.h [hunk 8]
```diff
@@ -109,4 +112,19 @@ typedef NS_ENUM(NSUInteger, MPPasswordEncoding) {
  */
 + (NSString *)defaultControllerPathForKey:(NSString *)key;
 
+/**
+ *  Returns the names of the databases with enabled Touch ID unlocking.
+ */
++ (NSArray*)getTouchIdEnabledDatabases;
```
Comment (mstarke) [inline MacPass/MPSettingsHelper.h]: simple getters do not have a get prefix.

**Comment by L4grange** [comment]
Good observations. If you have any ideas about the root UUID let me know, otherwise I'll have to keep using the DB name, which is not the best choice. 

About using Swift: I'm most familiar with Swift and still not that fluent with ObjC but I can understand your concerns. The KeychainPasswordItem was a class I got from [Apple's official examples](https://developer.apple.com/library/content/samplecode/GenericKeychain/Introduction/Intro.html) in order to get this working as soon as possible. It should not be too hard for me to translate it.

Thanks for your other comments, I'll do some cleanup. Keep them coming, I'll commit any changes I make later today :)

**Comment by mstarke** [comment]
@L4grange Any updates on the Objective-C conversion? I might try to do it myself if you're ok with it.

**Comment by L4grange** [comment]
Hey sorry for keeping quiet for so long. I've been very busy with work lately so I did not have the time of looking properly into this. I've made some progress but I got stuck at translating UnsafeMutablePointer to objc.

I could commit my unfinished file if you want, or you can do it yourself. Either way feel free to do it, since unfortunately I don't have the time right now.

**Comment by mstarke** [comment]
No worries. I'm in no position to criticise the development speed. If you want to share your progress, I'm happily trying to get it to work!

**Comment by caesar** [comment]
Great to see some progress here, thanks guys! I'm really looking forward to this.

Incidentally, Keypass2Android's approach here might be worth considering. It has two options for using the fingerprint: either allow using it to actually unlock the database, which requires storing the password ('securely'), or only use fingerprint for 'quick unlock' - meaning that once the database is unlocked, the app 'locks' after a certain amount of time (but the database is still open in memory), and switching back to the app requires using the fingerprint.
Might be interesting to have the same options in MacPass?

**Comment by 9point6** [comment]
@L4grange could you push up the unfinished work? How much more work do you think is needed?

**Comment by L4grange** [comment]
Hey guys, sorry for being gone. Unfortunatelly, I do not have the time to finish this, but I will try to push what I've done in the next days. Not much more work should be needed, just a translation of the KeychainPasswordItem.swift file in objc, and removing any Swift leftovers. Keep in mind that my branch is really old (now) and there will be conflicts. I'll let you know when I push my changes

**Comment by wizard86pz** [comment]
Thanks @L4grange. I was going to make a new pull request before to find your one. I can't wait to see first steps inside the codes and to help in case of translations in italian and english if needed.

**Comment by wizard86pz** [comment]
@L4grange no news about?

**Comment by L4grange** [comment]
I checked my code today and there is not much progress beside the translation of KeychainPasswordItem.swift file to objc, which is not complete. It's a small file and it should not be hard to translate. If anyone wants to take over this feature, please do, you can use my touchid_support branch on my fork as the base and rebase the project.

I use MacPass every day and I'm really sorry I could not put this feature out, it is something that would benefit myself as well, but life got in the way and I don't think I'll get the time to finish it. I would love to see someone making this work, the way its supposed to.

If I ever get find the free time, I'll try to start working on this again, but I can't promise you anything, so don't count on me. Good luck to anyone who tries, and thank to everyone who has contributed in making MacPass the beautiful password manager we use daily.

**Comment by georgesnow** [comment]
There is another open issue #514 about this. I mentioned over there I will take stab at it cause I don’t have Touch ID Mac.

**Comment by L4grange** [comment]
Thank you @georgesnow . I can test your builds and report, since my MacBook has Touch ID, so let me know of any progress.

**Comment by georgesnow** [comment]
@L4grange  alright well I have no idea if it will work. I used your code and converted it using SAMKeychain library, which uses objective-c. I can't quite tell how far you got on the functionally, but here is what I have, it builds with no errors so 🤷‍♂️ let me know: 
**_(definitely don't test on production keepass database)_**


https://github.com/georgesnow/MacPass/tree/touchid


if anyone knows how to test Touch ID on Mac that doesn't have it let me know. I could only find info for simulating Touch ID on iOS

**Comment by georgesnow** [comment]
I do know the UUID thing needs to get figured out and not sure if the NSUserDefaults will work.

**Comment by DominusExult** [comment]
If you have a build available I can test that on my wife's MacBook that has touchID

**Comment by georgesnow** [comment]
https://github.com/georgesnow/MacPass/releases/tag/CRUDEBuild-A

big disclaimer cause I have no idea what will and won't work. tread carefully.

**Comment by DominusExult** [comment]
Thanks, I'll definitely won't use it with my database ;)
It's nighttime now but tomorrow I'll give it a go

**Comment by georgesnow** [comment]
updated to support Touch ID (and I also included the ability to save your keychain). It uses the same prompts, but that should get cleaned up before merging.

https://github.com/georgesnow/MacPass/releases/tag/CrudeBuild-D

**Comment by georgesnow** [comment]
@L4grange - if you get a chance let me know if this build works for you.

**Comment by BrainBug0815** [comment]
@georgesnow First I want to thank you that you take up the work for that long awaited feature.
I am very curious testing your progress. 
Unfortunately I have issues using the current build CrudeBuild-D.
As soon as I use my fingerprint MacPass crashes.

I have downloaded the app and just started directly out of the download folder.

please let me know what detailed information you need.
for now, an excerpt from the crash report (the message changes each time):
Crashed Thread:        1  Dispatch queue: com.apple.root.default-qos

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Application Specific Information:
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'NSWindow drag regions should only be invalidated on the Main Thread!'
terminating with uncaught exception of type NSException
abort() called

Thread 1 Crashed:: Dispatch queue: com.apple.root.default-qos
0   libsystem_kernel.dylib        	0x00007fff624632c6 __pthread_kill + 10
1   libsystem_pthread.dylib       	0x00007fff6251ebf1 pthread_kill + 284
2   libsystem_c.dylib             	0x00007fff623cd6a6 abort + 127
3   libc++abi.dylib               	0x00007fff5f5a9641 abort_message + 231
4   libc++abi.dylib               	0x00007fff5f5a97df default_terminate_handler() + 267
5   libobjc.A.dylib               	0x00007fff60b5cee3 _objc_terminate() + 97
6   libc++abi.dylib               	0x00007fff5f5b519e std::__terminate(void (*)()) + 8
7   libc++abi.dylib               	0x00007fff5f5b5213 std::terminate() + 51
8   libdispatch.dylib             	0x00007fff622db651 _dispatch_client_callout + 28
9   libdispatch.dylib             	0x00007fff622dda53 _dispatch_queue_override_invoke + 763
10  libdispatch.dylib             	0x00007fff622e93bc _dispatch_root_queue_drain + 324
11  libdispatch.dylib             	0x00007fff622e9b46 _dispatch_worker_thread2 + 90
12  libsystem_pthread.dylib       	0x00007fff6251b6b3 _pthread_wqthread + 583
13  libsystem_pthread.dylib       	0x00007fff6251b3fd start_wqthread + 13

**Comment by georgesnow** [comment]
give the more recent build a try if you can:

https://github.com/georgesnow/MacPass/releases/tag/CrudeBuild-E

check out issue #514 that's where I have been trying to track the feedback

**Pull request closed without merging**